[ML] Fix race condition between job opening and feature reset

droberts195 opened the pull request:
There was a point during the job opening sequence where performing
a feature reset could hang.

This happened when the kill request issued by feature reset was
executed after the job's persistent task was assigned but before
the job's native process was started. The persistent task was
incorrectly left running in this situation, yet the job opening
sequence was aborted which meant the subsequent close request
issued by feature reset would wait for a very long time for the
persistent task to disappear.

The fix is to make the kill process request cancel the persistent
task consistently based on its request parameters and not on the
current state of the task.

Fixes #74141

droberts195 committed:
[ML] Fix race condition between job opening and feature reset

There was a point during the job opening sequence where performing
a feature reset could hang.

This happened when the kill request issued by feature reset was
executed after the job's persistent task was assigned but before
the job's native process was started. The persistent task was
incorrectly left running in this situation, yet the job opening
sequence was aborted which meant the subsequent close request
issued by feature reset would wait for a very long time for the
persistent task to disappear.

The fix is to make the kill process request cancel the persistent
task consistently based on its request parameters and not on the
current state of the task.

Fixes #74141

elasticmachine commented:
Pinging @elastic/ml-core (Team:ML)

droberts195 commented:
Although there are no explicit tests to prove this fix works, we do a feature reset at the end of _every integration test_, so it gets a lot of test coverage that way.

davidkyle commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/ProcessContext.java:
```diff
@@ -128,10 +132,27 @@ KillBuilder setShouldFinalizeJob(boolean shouldFinalizeJob) {
+
+        /**
+         * @param finish This argument overrides the member variable of the same name. Sometimes this method is called
+         *               immediately after a process is started, if it was requested to be killed before it started,
```
```suggestion
         *               immediately after a process is started, if it was requested to be killed before it started.
```

davidkyle commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/ProcessContext.java:
```diff
@@ -128,10 +132,27 @@ KillBuilder setShouldFinalizeJob(boolean shouldFinalizeJob) {
+        /**
+         * @param finish This argument overrides the member variable of the same name. Sometimes this method is called
+         *               immediately after a process is started, if it was requested to be killed before it started,
+         *               and in this situation the persistent task has already been completed if this was desired.
```
```suggestion
         *               In this situation the persistent task has already been completed if that was desired.
```

droberts195 committed:
Apply suggestions from code review

Co-authored-by: David Kyle <david.kyle@elastic.co>

droberts195 committed:
Merge branch 'master' into fix_upgrade_mode_test_again

droberts195 committed:
Rename variable

droberts195 committed:
Making debug messages more precise

droberts195 committed:
Remove redundant variable

droberts195 committed:
Merge branch 'master' into fix_upgrade_mode_test_again

droberts195 committed:
Simplification

The changes of #74415 made some of the changes of #71656
redundant. This commit is deleting code from #71656 that
would never execute now.

davidkyle reviewed (approved):
LGTM
